File: netcomponents_rfq/node/submit_rfqs.js
/**
 * Submit RFQs to qualifying in-stock suppliers
 *
 * Criteria:
 * - In-stock inventory only (skip brokered)
 * - Skip franchised/authorized distributors
 * - Skip Asia/Other region (separate purchasing group)
 * - Supplier qty must be >= requested qty (or take largest if none qualify)
 * - Max 3-4 suppliers per region
 * - Europe suppliers: add "Please confirm country of origin" message
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');
const config = require('./config');

const SCREENSHOTS_DIR = path.join(__dirname, 'screenshots');
if (!fs.existsSync(SCREENSHOTS_DIR)) fs.mkdirSync(SCREENSHOTS_DIR, { recursive: true });

const MAX_SUPPLIERS_PER_REGION = 3;
// Franchised/authorized distributors are identified by 'ncauth' class in DOM
// Independent distributors have 'ncnoauth' class - no hardcoded name list needed

async function delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function screenshot(page, name) {
    const filename = path.join(SCREENSHOTS_DIR, `rfq_${name}.png`);
    await page.screenshot({ path: filename, fullPage: false });
    console.log(`    Screenshot: rfq_${name}.png`);
}

async function main() {
    const partNumber = process.argv[2];
    const quantity = parseInt(process.argv[3], 10);

    if (!partNumber || !quantity) {
        console.log('Usage: node submit_rfqs.js <part_number> <quantity>');
        console.log('Example: node submit_rfqs.js "DS3231SN#" 1000');
        process.exit(1);
    }

    console.log('========================================');
    console.log('NetComponents RFQ Submission');
    console.log(`Part: ${partNumber}`);
    console.log(`Quantity: ${quantity}`);
    console.log(`Max suppliers per region: ${MAX_SUPPLIERS_PER_REGION}`);
    console.log('========================================\n');

    const startTime = Date.now();
    const timing = {
        login: 0,
        search: 0,
        suppliers: [],
        total: 0
    };

    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({ viewport: { width: 1400, height: 1000 } });
    const page = await context.newPage();

    const results = [];

    try {
        // 1. Login
        console.log('1. Logging in...');
        const loginStart = Date.now();
        await page.goto(config.BASE_URL);
        await delay(2000);
        await page.click('a:has-text("Login")');
        await delay(2000);
        await page.fill('#AccountNumber', config.NETCOMPONENTS_ACCOUNT);
        await page.fill('#UserName', config.NETCOMPONENTS_USERNAME);
        await page.fill('#Password', config.NETCOMPONENTS_PASSWORD);
        await page.press('#Password', 'Enter');
        await delay(5000);
        timing.login = (Date.now() - loginStart) / 1000;
        console.log(`   Done (${timing.login.toFixed(1)}s)\n`);

        // 2. Search for part
        console.log(`2. Searching for ${partNumber}...`);
        const searchStart = Date.now();
        await page.fill('#PartsSearched_0__PartNumber', partNumber);
        await page.click('#btnSearch');
        await delay(8000);
        timing.search = (Date.now() - searchStart) / 1000;
        console.log(`   Done (${timing.search.toFixed(1)}s)\n`);

        // 3. Parse all suppliers and aggregate by supplier name
        console.log('3. Finding qualifying suppliers...');
        const rows = await page.$$('table#trv_0 tbody tr');

        const supplierData = {};  // key: "name|region", value: {name, region, totalQty, rows: [rowElements]}
        let inStockSection = false;
        let currentRegion = 'Unknown';

        for (const row of rows) {
            const rowText = (await row.innerText().catch(() => '')).toLowerCase();

            // Track region headers
            if (rowText.includes('americas') && !rowText.includes('inventory') && rowText.length < 50) {
                currentRegion = 'Americas';
                continue;
            }
            if (rowText.includes('europe') && !rowText.includes('inventory') && rowText.length < 50) {
                currentRegion = 'Europe';
                continue;
            }
            if ((rowText.includes('asia') || rowText.includes('other')) && !rowText.includes('inventory') && rowText.length < 50) {
                currentRegion = 'Asia/Other';
                continue;
            }

            // Check for section subheader rows
            if ((rowText.startsWith('in stock') || rowText.startsWith('in-stock')) && rowText.length < 100) {
                inStockSection = true;
                continue;
            }
            if ((rowText.startsWith('brokered inventory') || rowText.startsWith('brokered')) && rowText.length < 100) {
                inStockSection = false;
                continue;
            }

            // Skip if not in-stock section or Asia/Other
            if (!inStockSection) continue;
            if (currentRegion === 'Asia/Other') continue;

            const cells = await row.$$('td');
            if (cells.length < 16) continue;

            // Get supplier name
            const supplierCell = cells[15];
            const link = await supplierCell.$('a');
            if (!link) continue;
            const supplierName = (await link.innerText()).trim();
            if (!supplierName) continue;

            // Skip franchised/authorized distributors (marked with 'ncauth' class)
            const authIcon = await supplierCell.$('.ncauth');
            if (authIcon) continue;

            // Get quantity from column 8
            let qty = 0;
            try {
                const qtyText = (await cells[8].innerText()).trim();
                const qtyMatch = qtyText.replace(/,/g, '').match(/^(\d+)/);
                if (qtyMatch) qty = parseInt(qtyMatch[1], 10);
            } catch (e) {}

            // Aggregate by supplier
            const key = `${supplierName}|${currentRegion}`;
            if (!supplierData[key]) {
                supplierData[key] = { name: supplierName, region: currentRegion, totalQty: 0, link: link };
            }
            supplierData[key].totalQty += qty;
            // Keep the link with highest qty
            if (qty > 0) {
                supplierData[key].link = link;
            }
        }

        // Convert to array and sort by qty
        const allSuppliers = Object.values(supplierData).sort((a, b) => b.totalQty - a.totalQty);

        // Split by region and filter by qty
        const americas = allSuppliers.filter(s => s.region === 'Americas');
        const europe = allSuppliers.filter(s => s.region === 'Europe');

        // Select suppliers: prefer those meeting qty, fallback to largest
        const americasMeetQty = americas.filter(s => s.totalQty >= quantity);
        const europeMeetQty = europe.filter(s => s.totalQty >= quantity);

        const selectedAmericas = americasMeetQty.length > 0
            ? americasMeetQty.slice(0, MAX_SUPPLIERS_PER_REGION)
            : americas.slice(0, MAX_SUPPLIERS_PER_REGION);

        const selectedEurope = europeMeetQty.length > 0
            ? europeMeetQty.slice(0, MAX_SUPPLIERS_PER_REGION)
            : europe.slice(0, MAX_SUPPLIERS_PER_REGION);

        console.log(`   Americas: ${selectedAmericas.length} suppliers selected`);
        selectedAmericas.forEach(s => console.log(`     - ${s.name} (${s.totalQty.toLocaleString()})`));
        console.log(`   Europe: ${selectedEurope.length} suppliers selected`);
        selectedEurope.forEach(s => console.log(`     - ${s.name} (${s.totalQty.toLocaleString()})`));
        console.log('');

        const allSelected = [...selectedAmericas, ...selectedEurope];

        if (allSelected.length === 0) {
            console.log('   No qualifying suppliers found!');
            return;
        }

        // 4. Submit RFQs to each supplier
        console.log(`4. Submitting RFQs to ${allSelected.length} suppliers...\n`);

        for (let i = 0; i < allSelected.length; i++) {
            const supplier = allSelected[i];
            const supplierStart = Date.now();
            console.log(`   [${i + 1}/${allSelected.length}] ${supplier.name} (${supplier.region})...`);

            try {
                // Re-do search to get fresh page state
                await page.goto(config.BASE_URL);
                await delay(2000);
                await page.fill('#PartsSearched_0__PartNumber', partNumber);
                await page.click('#btnSearch');
                await delay(6000);

                // Find and click the supplier
                const supplierLink = await page.$(`a:has-text("${supplier.name}")`);
                if (!supplierLink) {
                    console.log(`    ERROR: Could not find supplier link`);
                    results.push({ supplier: supplier.name, region: supplier.region, status: 'FAILED', error: 'Supplier not found' });
                    continue;
                }

                await supplierLink.click();
                await delay(2000);

                // Click E-Mail RFQ link
                const rfqLink = await page.$('a:has-text("E-Mail RFQ")');
                if (!rfqLink) {
                    console.log(`    ERROR: E-Mail RFQ option not available`);
                    results.push({ supplier: supplier.name, region: supplier.region, status: 'FAILED', error: 'No RFQ option' });
                    // Close panel if open
                    await page.keyboard.press('Escape');
                    await delay(1000);
                    continue;
                }

                await rfqLink.click();
                await delay(2000);

                // Fill the RFQ form
                // Wait for form to load
                await delay(1000);

                // Check part checkbox
                const partCheckbox = await page.$('#Parts_0__Selected');
                if (partCheckbox) {
                    const isChecked = await partCheckbox.isChecked();
                    if (!isChecked) {
                        await partCheckbox.check();
                        console.log(`    Checked part selection`);
                    }
                } else {
                    console.log(`    WARNING: Part checkbox not found`);
                }

                // Fill quantity - try multiple selectors
                let qtyInput = await page.$('#Parts_0__Quantity');
                if (!qtyInput) {
                    qtyInput = await page.$('input[name="Parts[0].Quantity"]');
                }
                if (!qtyInput) {
                    // Try to find any input in the Qty Requested column area
                    qtyInput = await page.$('input[type="text"][placeholder*="Qty"]');
                }
                if (!qtyInput) {
                    // Look for input near "Qty Requested" text
                    const inputs = await page.$$('input[type="text"]');
                    for (const inp of inputs) {
                        const name = await inp.getAttribute('name');
                        const id = await inp.getAttribute('id');
                        if ((name && name.toLowerCase().includes('quantity')) ||
                            (id && id.toLowerCase().includes('quantity'))) {
                            qtyInput = inp;
                            break;
                        }
                    }
                }

                if (qtyInput) {
                    await qtyInput.click();
                    await qtyInput.fill(quantity.toString());
                    console.log(`    Entered quantity: ${quantity}`);
                } else {
                    console.log(`    WARNING: Quantity input not found`);
                }

                // Add Europe message
                if (supplier.region === 'Europe') {
                    let commentsField = await page.$('#Comments');
                    if (!commentsField) {
                        commentsField = await page.$('textarea[name="Comments"]');
                    }
                    if (!commentsField) {
                        commentsField = await page.$('textarea');
                    }
                    if (commentsField) {
                        await commentsField.fill('Please confirm country of origin.');
                        console.log(`    Added Europe COO message`);
                    }
                }

                await delay(1000);
                await screenshot(page, `${i + 1}_form_filled`);

                // Find Send RFQ button - it's an INPUT type="button"
                let sendBtn = await page.$('input[type="button"].action-btn');
                if (!sendBtn) {
                    sendBtn = await page.$('input[value="Send RFQ"]');
                }
                if (!sendBtn) {
                    sendBtn = await page.$('input.btn-primary[type="button"]');
                }

                if (sendBtn) {
                    const isDisabled = await sendBtn.getAttribute('disabled');
                    const btnText = (await sendBtn.innerText().catch(() => '')).trim();
                    console.log(`    Found button: "${btnText}" disabled=${isDisabled !== null}`);

                    if (isDisabled === null) {
                        await sendBtn.click();
                        await delay(3000);
                        await screenshot(page, `${i + 1}_after_send`);
                        const supplierTime = (Date.now() - supplierStart) / 1000;
                        console.log(`    SUCCESS: RFQ sent (${supplierTime.toFixed(1)}s)`);
                        timing.suppliers.push({ name: supplier.name, time: supplierTime, status: 'SENT' });
                        results.push({
                            supplier: supplier.name,
                            region: supplier.region,
                            qty: quantity,
                            status: 'SENT',
                            timestamp: new Date().toISOString()
                        });
                    } else {
                        console.log(`    ERROR: Send button disabled`);
                        await screenshot(page, `${i + 1}_disabled`);
                        results.push({ supplier: supplier.name, region: supplier.region, status: 'FAILED', error: 'Send button disabled' });
                    }
                } else {
                    console.log(`    ERROR: Send RFQ button not found`);
                    await screenshot(page, `${i + 1}_no_button`);
                    results.push({ supplier: supplier.name, region: supplier.region, status: 'FAILED', error: 'No Send button' });
                }

                // Close the form/modal
                await page.keyboard.press('Escape');
                await delay(1000);

            } catch (error) {
                const supplierTime = (Date.now() - supplierStart) / 1000;
                console.log(`    ERROR: ${error.message} (${supplierTime.toFixed(1)}s)`);
                timing.suppliers.push({ name: supplier.name, time: supplierTime, status: 'FAILED' });
                results.push({ supplier: supplier.name, region: supplier.region, status: 'FAILED', error: error.message });
            }
        }

        // 5. Summary
        timing.total = (Date.now() - startTime) / 1000;
        const avgPerSupplier = timing.suppliers.length > 0
            ? timing.suppliers.reduce((sum, s) => sum + s.time, 0) / timing.suppliers.length
            : 0;

        console.log('\n========================================');
        console.log('RFQ SUBMISSION SUMMARY');
        console.log('========================================');
        console.log(`Part: ${partNumber}`);
        console.log(`Quantity: ${quantity}`);
        console.log(`Total submitted: ${results.filter(r => r.status === 'SENT').length}/${results.length}`);
        console.log('');
        results.forEach(r => {
            const status = r.status === 'SENT' ? '✓' : '✗';
            const supplierTiming = timing.suppliers.find(s => s.name === r.supplier);
            const timeStr = supplierTiming ? ` (${supplierTiming.time.toFixed(1)}s)` : '';
            console.log(`${status} ${r.supplier} (${r.region}) - ${r.status}${timeStr}${r.error ? ': ' + r.error : ''}`);
        });

        console.log('\n========================================');
        console.log('TIMING');
        console.log('========================================');
        console.log(`Login:              ${timing.login.toFixed(1)}s`);
        console.log(`Initial search:     ${timing.search.toFixed(1)}s`);
        console.log(`Avg per supplier:   ${avgPerSupplier.toFixed(1)}s`);
        console.log(`Total runtime:      ${timing.total.toFixed(1)}s (${(timing.total / 60).toFixed(1)} min)`);
        console.log(`Suppliers/minute:   ${(timing.suppliers.length / (timing.total / 60)).toFixed(1)}`);

    } catch (error) {
        console.error('\nFATAL ERROR:', error.message);
        await screenshot(page, 'error');
    } finally {
        await browser.close();
    }
}

main().catch(console.error);

File: netcomponents_rfq/python/submit_rfqs.py
#!/usr/bin/env python3
"""
Submit RFQs to qualifying in-stock suppliers

Criteria:
- In-stock inventory only (skip brokered)
- Skip franchised/authorized distributors
- Skip Asia/Other region (separate purchasing group)
- Supplier qty must be >= requested qty (or take largest if none qualify)
- Max 3 suppliers per region
- Europe suppliers: add "Please confirm country of origin" message

Usage:
    python submit_rfqs.py <part_number> <quantity>

Example:
    python submit_rfqs.py "DS3231SN#" 1000
"""

import sys
import asyncio
import time
import re
from datetime import datetime
from playwright.async_api import async_playwright
import config


async def screenshot(page, name):
    """Save a screenshot"""
    filename = config.SCREENSHOTS_DIR / f'rfq_{name}.png'
    await page.screenshot(path=str(filename), full_page=False)
    print(f'    Screenshot: rfq_{name}.png')


async def main():
    if len(sys.argv) < 3:
        print('Usage: python submit_rfqs.py <part_number> <quantity>')
        print('Example: python submit_rfqs.py "DS3231SN#" 1000')
        sys.exit(1)

    part_number = sys.argv[1]
    quantity = int(sys.argv[2])

    print('=' * 40)
    print('NetComponents RFQ Submission')
    print(f'Part: {part_number}')
    print(f'Quantity: {quantity:,}')
    print(f'Max suppliers per region: {config.MAX_SUPPLIERS_PER_REGION}')
    print('=' * 40)
    print()

    start_time = time.time()
    timing = {
        'login': 0,
        'search': 0,
        'suppliers': [],
        'total': 0
    }

    results = []
    omitted_suppliers = []  # Track suppliers filtered out by min order value

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context(viewport={'width': 1400, 'height': 1000})
        page = await context.new_page()

        try:
            # 1. Login
            print('1. Logging in...')
            login_start = time.time()
            await page.goto(config.BASE_URL)
            await asyncio.sleep(2)
            await page.click('a:has-text("Login")')
            await asyncio.sleep(2)
            await page.fill('#AccountNumber', config.NETCOMPONENTS_ACCOUNT)
            await page.fill('#UserName', config.NETCOMPONENTS_USERNAME)
            await page.fill('#Password', config.NETCOMPONENTS_PASSWORD)
            await page.press('#Password', 'Enter')
            await asyncio.sleep(5)
            timing['login'] = time.time() - login_start
            print(f"   Done ({timing['login']:.1f}s)\n")

            # 2. Search for part
            print(f'2. Searching for {part_number}...')
            search_start = time.time()
            await page.fill('#PartsSearched_0__PartNumber', part_number)
            await page.click('#btnSearch')
            await asyncio.sleep(8)
            timing['search'] = time.time() - search_start
            print(f"   Done ({timing['search']:.1f}s)\n")

            # 3. Parse all suppliers and aggregate by supplier name
            print('3. Finding qualifying suppliers...')
            rows = await page.query_selector_all('table#trv_0 tbody tr')

            supplier_data = {}  # key: "name|region" -> {name, region, total_qty, link}
            in_stock_section = False
            current_region = 'Unknown'

            for row in rows:
                row_text = (await row.inner_text() or '').lower()

                # Track region headers
                if 'americas' in row_text and 'inventory' not in row_text and len(row_text) < 50:
                    current_region = 'Americas'
                    continue
                if 'europe' in row_text and 'inventory' not in row_text and len(row_text) < 50:
                    current_region = 'Europe'
                    continue
                if ('asia' in row_text or 'other' in row_text) and 'inventory' not in row_text and len(row_text) < 50:
                    current_region = 'Asia/Other'
                    continue

                # Check for section subheader rows
                if (row_text.startswith('in stock') or row_text.startswith('in-stock')) and len(row_text) < 100:
                    in_stock_section = True
                    continue
                if (row_text.startswith('brokered inventory') or row_text.startswith('brokered')) and len(row_text) < 100:
                    in_stock_section = False
                    continue

                # Skip if not in-stock or Asia/Other
                if not in_stock_section:
                    continue
                if current_region == 'Asia/Other':
                    continue

                cells = await row.query_selector_all('td')
                if len(cells) < 16:
                    continue

                # Get supplier name from column 15
                supplier_cell = cells[15]
                link = await supplier_cell.query_selector('a')
                if not link:
                    continue
                supplier_name = (await link.inner_text()).strip()
                if not supplier_name:
                    continue

                # Skip franchised/authorized distributors (marked with 'ncauth' class)
                auth_icon = await supplier_cell.query_selector('.ncauth')
                if auth_icon:
                    continue

                # Get date code from column 4
                dc_text = ''
                dc_year = None
                dc_ambiguous = False
                try:
                    dc_text = (await cells[4].inner_text()).strip()
                    dc_year, dc_ambiguous = config.parse_date_code(dc_text)
                except Exception:
                    pass

                # Get quantity from column 8
                qty = 0
                try:
                    qty_text = (await cells[8].inner_text()).strip()
                    qty_clean = qty_text.replace(',', '')
                    match = re.match(r'^(\d+)', qty_clean)
                    if match:
                        qty = int(match.group(1))
                except Exception:
                    pass

                # Aggregate by supplier
                key = f"{supplier_name}|{current_region}"
                if key not in supplier_data:
                    supplier_data[key] = {
                        'name': supplier_name,
                        'region': current_region,
                        'total_qty': 0,
                        'link': link,
                        'best_dc_year': None,
                        'best_dc_text': '',
                        'dc_ambiguous': False
                    }
                supplier_data[key]['total_qty'] += qty

                # Keep the best (freshest) date code
                if dc_year is not None:
                    if supplier_data[key]['best_dc_year'] is None or dc_year > supplier_data[key]['best_dc_year']:
                        supplier_data[key]['best_dc_year'] = dc_year
                        supplier_data[key]['best_dc_text'] = dc_text
                        supplier_data[key]['dc_ambiguous'] = dc_ambiguous

                # Keep the link with highest qty
                if qty > 0:
                    supplier_data[key]['link'] = link

            # Determine date code status for each supplier
            for s in supplier_data.values():
                s['dc_status'] = config.get_dc_status(s.get('best_dc_year'), s.get('dc_ambiguous', False))

            # Split by region
            americas = [s for s in supplier_data.values() if s['region'] == 'Americas']
            europe = [s for s in supplier_data.values() if s['region'] == 'Europe']

            # Sort by priority score (fresh DC + qty prioritized, unknown DC given benefit of doubt)
            americas.sort(key=lambda x: config.supplier_priority_score(x, quantity), reverse=True)
            europe.sort(key=lambda x: config.supplier_priority_score(x, quantity), reverse=True)

            # Apply coverage-based filtering to remove tiny-qty suppliers when good coverage exists
            all_suppliers = americas + europe
            filtered_all = config.filter_by_coverage(all_suppliers, quantity)
            filtered_names = {s['name'] for s in filtered_all}

            americas = [s for s in americas if s['name'] in filtered_names]
            europe = [s for s in europe if s['name'] in filtered_names]

            # Use cross-region balancing: if one region is short, give extra slots to the other
            americas_slots, europe_slots = config.calculate_region_slots(len(americas), len(europe))

            selected_americas = americas[:americas_slots]
            selected_europe = europe[:europe_slots]

            # Add +1 extra if unknown DCs present (buffer for uncertainty)
            if config.should_add_extra_supplier(selected_americas) and len(americas) > americas_slots:
                selected_americas = americas[:americas_slots + 1]
            if config.should_add_extra_supplier(selected_europe) and len(europe) > europe_slots:
                selected_europe = europe[:europe_slots + 1]

            print(f'   Americas: {len(selected_americas)} suppliers selected')
            for s in selected_americas:
                dc_info = f", DC:{s['best_dc_text']}" if s.get('best_dc_text') else " (no DC)"
                status = f" [{s['dc_status'].upper()}]" if s.get('dc_status') else ""
                print(f"     - {s['name']} ({s['total_qty']:,}{dc_info}){status}")
            print(f'   Europe: {len(selected_europe)} suppliers selected')
            for s in selected_europe:
                dc_info = f", DC:{s['best_dc_text']}" if s.get('best_dc_text') else " (no DC)"
                status = f" [{s['dc_status'].upper()}]" if s.get('dc_status') else ""
                print(f"     - {s['name']} ({s['total_qty']:,}{dc_info}){status}")
            print()

            all_selected = selected_americas + selected_europe

            if not all_selected:
                print('   No qualifying suppliers found!')
                return

            # 4. Submit RFQs to each supplier
            print(f'4. Submitting RFQs to {len(all_selected)} suppliers...\n')

            for i, supplier in enumerate(all_selected):
                supplier_start = time.time()

                # Adjust quantity if supplier has less than requested
                rfq_qty, qty_adjusted = config.adjust_rfq_quantity(quantity, supplier['total_qty'])
                qty_note = f" (adjusted from {quantity})" if qty_adjusted else ""

                print(f"   [{i + 1}/{len(all_selected)}] {supplier['name']} ({supplier['region']})...")

                try:
                    # Re-do search to get fresh page state
                    await page.goto(config.BASE_URL)
                    await asyncio.sleep(2)
                    await page.fill('#PartsSearched_0__PartNumber', part_number)
                    await page.click('#btnSearch')
                    await asyncio.sleep(6)

                    # Find and click the supplier
                    supplier_link = await page.query_selector(f'a:has-text("{supplier["name"]}")')
                    if not supplier_link:
                        print('    ERROR: Could not find supplier link')
                        results.append({
                            'supplier': supplier['name'],
                            'region': supplier['region'],
                            'status': 'FAILED',
                            'error': 'Supplier not found'
                        })
                        continue

                    await supplier_link.click()
                    await asyncio.sleep(3)  # Wait for supplier detail popup

                    # Extract min order value from supplier detail popup
                    min_order_value = await config.extract_min_order_value(page)
                    supplier['min_order_value'] = min_order_value
                    if min_order_value:
                        print(f'    Min order value: ${min_order_value:.2f}')

                    # TODO: Apply min order value filter when franchise_data is available
                    # This will be integrated when called from batch_rfqs_from_system.py
                    # with franchise screening data

                    # Click E-Mail RFQ link
                    rfq_link = await page.query_selector('a:has-text("E-Mail RFQ")')
                    if not rfq_link:
                        print('    ERROR: E-Mail RFQ option not available')
                        results.append({
                            'supplier': supplier['name'],
                            'region': supplier['region'],
                            'status': 'FAILED',
                            'error': 'No RFQ option'
                        })
                        await page.keyboard.press('Escape')
                        await asyncio.sleep(1)
                        continue

                    await rfq_link.click()
                    await asyncio.sleep(2)

                    # Fill the RFQ form
                    await asyncio.sleep(1)

                    # Check part checkbox
                    part_checkbox = await page.query_selector('#Parts_0__Selected')
                    if part_checkbox:
                        is_checked = await part_checkbox.is_checked()
                        if not is_checked:
                            await part_checkbox.check()
                            print('    Checked part selection')
                    else:
                        print('    WARNING: Part checkbox not found')

                    # Fill quantity
                    qty_input = await page.query_selector('#Parts_0__Quantity')
                    if not qty_input:
                        qty_input = await page.query_selector('input[name="Parts[0].Quantity"]')
                    if not qty_input:
                        qty_input = await page.query_selector('input[type="text"][placeholder*="Qty"]')
                    if not qty_input:
                        inputs = await page.query_selector_all('input[type="text"]')
                        for inp in inputs:
                            name = await inp.get_attribute('name')
                            id_attr = await inp.get_attribute('id')
                            if ((name and 'quantity' in name.lower()) or
                                (id_attr and 'quantity' in id_attr.lower())):
                                qty_input = inp
                                break

                    if qty_input:
                        await qty_input.click()
                        await qty_input.fill(str(rfq_qty))
                        print(f'    Entered quantity: {rfq_qty}{qty_note}')
                    else:
                        print('    WARNING: Quantity input not found')

                    # Add Europe message
                    if supplier['region'] == 'Europe':
                        comments_field = await page.query_selector('#Comments')
                        if not comments_field:
                            comments_field = await page.query_selector('textarea[name="Comments"]')
                        if not comments_field:
                            comments_field = await page.query_selector('textarea')
                        if comments_field:
                            await comments_field.fill('Please confirm country of origin.')
                            print('    Added Europe COO message')

                    await asyncio.sleep(1)
                    await screenshot(page, f'{i + 1}_form_filled')

                    # Find Send RFQ button - it's an INPUT type="button"
                    send_btn = await page.query_selector('input[type="button"].action-btn')
                    if not send_btn:
                        send_btn = await page.query_selector('input[value="Send RFQ"]')
                    if not send_btn:
                        send_btn = await page.query_selector('input.btn-primary[type="button"]')

                    if send_btn:
                        is_disabled = await send_btn.get_attribute('disabled')
                        btn_text = ''
                        try:
                            btn_text = (await send_btn.inner_text()).strip()
                        except Exception:
                            pass
                        print(f'    Found button: "{btn_text}" disabled={is_disabled is not None}')

                        if is_disabled is None:
                            await send_btn.click()
                            await asyncio.sleep(3)
                            await screenshot(page, f'{i + 1}_after_send')
                            supplier_time = time.time() - supplier_start
                            print(f'    SUCCESS: RFQ sent ({supplier_time:.1f}s)')
                            timing['suppliers'].append({
                                'name': supplier['name'],
                                'time': supplier_time,
                                'status': 'SENT'
                            })
                            results.append({
                                'supplier': supplier['name'],
                                'region': supplier['region'],
                                'qty': quantity,
                                'status': 'SENT',
                                'timestamp': datetime.now().isoformat()
                            })
                        else:
                            print('    ERROR: Send button disabled')
                            await screenshot(page, f'{i + 1}_disabled')
                            results.append({
                                'supplier': supplier['name'],
                                'region': supplier['region'],
                                'status': 'FAILED',
                                'error': 'Send button disabled'
                            })
                    else:
                        print('    ERROR: Send RFQ button not found')
                        await screenshot(page, f'{i + 1}_no_button')
                        results.append({
                            'supplier': supplier['name'],
                            'region': supplier['region'],
                            'status': 'FAILED',
                            'error': 'No Send button'
                        })

                    # Close the form/modal
                    await page.keyboard.press('Escape')
                    await asyncio.sleep(1)

                except Exception as e:
                    supplier_time = time.time() - supplier_start
                    print(f'    ERROR: {e} ({supplier_time:.1f}s)')
                    timing['suppliers'].append({
                        'name': supplier['name'],
                        'time': supplier_time,
                        'status': 'FAILED'
                    })
                    results.append({
                        'supplier': supplier['name'],
                        'region': supplier['region'],
                        'status': 'FAILED',
                        'error': str(e)
                    })

            # 5. Summary
            timing['total'] = time.time() - start_time
            avg_per_supplier = (sum(s['time'] for s in timing['suppliers']) / len(timing['suppliers'])
                               if timing['suppliers'] else 0)

            print()
            print('=' * 40)
            print('RFQ SUBMISSION SUMMARY')
            print('=' * 40)
            print(f'Part: {part_number}')
            print(f'Quantity: {quantity:,}')
            sent_count = len([r for r in results if r['status'] == 'SENT'])
            print(f'Total submitted: {sent_count}/{len(results)}')
            print()

            for r in results:
                status = '✓' if r['status'] == 'SENT' else '✗'
                supplier_timing = next((s for s in timing['suppliers'] if s['name'] == r['supplier']), None)
                time_str = f" ({supplier_timing['time']:.1f}s)" if supplier_timing else ''
                error_str = f": {r.get('error', '')}" if r.get('error') else ''
                print(f"{status} {r['supplier']} ({r['region']}) - {r['status']}{time_str}{error_str}")

            print()
            print('=' * 40)
            print('TIMING')
            print('=' * 40)
            print(f"Login:              {timing['login']:.1f}s")
            print(f"Initial search:     {timing['search']:.1f}s")
            print(f"Avg per supplier:   {avg_per_supplier:.1f}s")
            print(f"Total runtime:      {timing['total']:.1f}s ({timing['total'] / 60:.1f} min)")
            if timing['total'] > 0:
                print(f"Suppliers/minute:   {len(timing['suppliers']) / (timing['total'] / 60):.1f}")

        except Exception as e:
            print(f'\nFATAL ERROR: {e}')
            await screenshot(page, 'error')
            import traceback
            traceback.print_exc()
        finally:
            await browser.close()


if __name__ == '__main__':
    asyncio.run(main())
